### PR TITLE
[#124] Some TZIP-17 Updates 

### LINKED ISSUE #124: Update TZIP-17 implementation
# Clarification and motivation

In #93, we implemented TZIP-17 in Stablecoin. 
Since then, we submitted some feedback [here](https://gitlab.com/tzip/tzip/-/merge_requests/83), and the TZIP has been revised.

For example:
* the TZIP now states that when a permit is issued twice, the contract should fail with `"DUP_PERMIT"`.
* the `revoke` entrypoint has been removed
* TZIP-16 compliance is now required
* `setExpiry`'s parameter type has changed
* a user can issue a permit for the `setExpiry` entrypoint
* there should be an upper bound limit on how big an expiry can be (e.g. 10^10 ~ 3000 years)

We should update our implementation to reflect these changes.

# Acceptance criteria

Stablecoin is TZIP-17-compliant.

### PR DESCRIPTION
**Pull request by sras**
## Description

Implements 

* Change to setExpiry's parameter type.
* Enable permits for the setExpiry entrypoint
* Set an upper bound limit on how big an expiry can be.

in additon, this also adds sender check for `Set_expiry` calls to check that the sender is the issuer of permit (Or a permit exists for the call), and the changes that revokes permit when expiry is set to zero.

There are some pending stuff.

1. One SMT test is failing during some test runs. EDIT: Has been fixed

 stack build --fast --test --test-arguments  '--hedgehog-replay "Size 4 Seed 4906785565932219612 2048738257461261607"'

There was a second hedgehog failure where this got triggered https://github.com/tqtezos/stablecoin/blob/master/haskell/test/SMT.hs#L659. I didn't save the replay parameter (Thought it was the above failure), and was not able to repro this fail after many runs.

2. Documetation updates. EDIT: Added

## Related issue(s)

Resolves #124 (Partial)

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [ ] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [ ] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [ ] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [ ] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: ligo/stablecoin/actions.ligo
Comment (gromakovsky) [inline ligo/stablecoin/actions.ligo:796]: `s/bytes/blake2b_hash`

File: haskell/src/Lorentz/Contracts/Stablecoin.hs
Comment (gromakovsky) [inline haskell/src/Lorentz/Contracts/Stablecoin.hs:129]: Why not have 3 fields here?

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (gromakovsky) [inline haskell/test/Lorentz/Contracts/Test/Permit.hs]: Is it supposed to be removed?

File: ligo/stablecoin/fa1.2/spec.ligo
Comment (gromakovsky) [inline ligo/stablecoin/fa1.2/spec.ligo:146]: Can we avoid having this constant in more than one place?

**Review by gromakovsky — CHANGES_REQUESTED** [review]
Mostly looks fine, have some minor comments. Can you please describe changes in commit messages before review, not after? These explanations actually make review easier.

Also this PR should definitely be reviewed by @dcastro as well since he wrote some code changed here.

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs]: How did you come across this? AFAIU, it is not in the final commit.

Comment (gromakovsky) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs]: Hm, right, seems to be removed.

> How did you come across this?

I was reviewing per-commit :shrug:

Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs]: >I was reviewing per-commit

Ah, makes sense. I ll keep it in mind. Do you want to rewrite commit history with more descriptive messages in this PR?

File: haskell/src/Lorentz/Contracts/Stablecoin.hs
Comment (sras) [reply haskell/src/Lorentz/Contracts/Stablecoin.hs]: Right. Changed.

File: ligo/stablecoin/actions.ligo
Comment (sras) [reply ligo/stablecoin/actions.ligo]: Changed.

File: ligo/stablecoin/fa1.2/spec.ligo
Comment (sras) [reply ligo/stablecoin/fa1.2/spec.ligo]: Changed

**Review by gromakovsky — APPROVED** [review]
> Do you want to rewrite commit history with more descriptive messages in this PR?

Yes, it should be done. Ideally, when I review a PR, I want to read a story explaining what happens in its commits, so that I don't have to go somewhere else and recall what needs to be done and why.
You can make @dcastro's review a bit easier by cleaning up commit history :) And please don't merge without his approval.

**Comment by sras** [comment]
@gromakovsky I have expanded the commit messages. Will wait for @dcastro to review.

File: ligo/stablecoin/permit.ligo
Comment (dcastro) [inline ligo/stablecoin/permit.ligo:249]: Can you please add this new error to the error table in the spec?

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (dcastro) [inline haskell/test/Lorentz/Contracts/Test/Permit.hs:309]: Can you please add a test: "setting the expiry of an already expired permit does not long prolong its life"

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (dcastro) [inline haskell/test/Lorentz/Contracts/Test/Permit.hs:386]: If I understand this correctly, I believe this can be simplified by using `withSender testPauser` and deleting the call to `callPermit ... Set_expiry`

```suggestion
            withSender testPauser $ do
              pausePermitHash <- callPermit stablecoinContract testPauserPK testPauserSK 0 Pause
              lCallEP stablecoinContract (Call @"Set_expiry") $
                SetExpiryParam testPauser 0  $ Just pausePermitHash
            lCallEP stablecoinContract (Call @"Pause") () `catchExpectedError` mgmNotPauser
```

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (dcastro) [inline haskell/test/Lorentz/Contracts/Test/Permit.hs:375]: This does not necessarily proof that the expiry was successfully set.

I think this should work:
1. Setting the default expiry to 1
    ```
        let defaultExpiry = 1
        let originationParams = defaultOriginationParams { opDefaultExpiry = defaultExpiry }
        withOriginated originate originationParams $ \stablecoinContract -> do
    ```
2. issue a permit for `SetExpiry` and then call `SetExpiry` with expiry=5
3. advance time by 3 seconds
4. consume permit by calling `Pause`

**Review by dcastro — CHANGES_REQUESTED** [review]
LGTM, just a couple of small notes.

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs:386]: @dcastro But the `testPauser` can call the `pause` entrypoint without requiring a permit, right?

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs:378]: Done.

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (dcastro) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs:389]: That's right. In my suggested change, the call to `pause` is sent from `nettest`, not from `testPauser`. Maybe it's not easy to spot here on github, but the call to `pause` is not inside the `withSender testPauser` block.

So it should still attempt to find a permit, and fail because we've set expiry=0

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs:309]: @dcastro This check was missing in my original changes. I have added it and the test for it now.

File: haskell/test/Lorentz/Contracts/Test/Permit.hs
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/Permit.hs]: Right.  Changed.

**Comment by dcastro** [comment]
Quick note: when you rebase on top of my changes, please:

1. Add `EXPIRY_TOO_BIG` to the list of errors in `haskell/src/Lorentz/Contracts/Stablecoin.hs` and `haskell/test/Lorentz/Contracts/Test/TZIP16.hs`, with an appropriate error message
2. Please amend the error description for `NOT_PERMIT_ISSUER` in the spec:
    ```
    // before
    The sender tried to revoke or set expiry on a permit
    // after
    The sender tried to set the expiry of a permit
    ```

File: docs/specification.md
Comment (gromakovsky) [inline docs/specification.md:659]: Looks like bad rebase.

File: haskell/package.yaml
Comment (gromakovsky) [inline haskell/package.yaml:7]: Please also update the changelog:
1. Record your changes there.
2. Update version to 1.4.0.

File: docs/specification.md
Comment (sras) [reply docs/specification.md:659]: Damn.

**Review by dcastro — APPROVED** [review]

**Pull request merged**
